Create rule S6566: Use DateTimeOffset instead of DateTime (#1717)
SonarSource/rspec: rules/S6566/csharp/rule.adoc

This rule recommends using `DateTimeOffset` instead of `DateTime` for projects targeting .NET Framework 2.0 or later.

== Why is this an issue?

> The `DateTimeOffset` type includes all the functionality of the `DateTime` type and also includes the time offset from Coordinated Universal Time (UTC). It is the preferred alternative to the `DateTime` object for avoiding many of the potential problems, without relying on third-party solutions.

Reviewer: Here I find the `potential problems` very generic. 
If we want to educate our users we need to give examples. :/

Reviewer 2: There is a Pitfalls chapter that provides a couple of examples of those problems, maybe I can add a (see "Pitfalls" chapter reference), wdyt?

> The `DateTimeOffset` type includes all the functionality of the `DateTime` type and also includes the time offset from Coordinated Universal Time (UTC). It is the preferred alternative to the `DateTime` object for avoiding many of the potential problems, without relying on third-party solutions.
> `DateTimeOffset` represents a point in time, typically expressed as a date and time of day, relative to UTC. It provides a greater degree of time zone awareness than the `DateTime` structure.

Reviewer: ```suggestion
You should use `DateTimeOffset` instead of `DateTime` as it provides all the information that the `DateTime` struct has, including the `Offset` property that defines the difference between the current DateTimeOffset instance's date and time and Coordinated Universal Time (UTC).
This way you can avoid potential problems created by the lack of timezone awareness (e.g., ambiguous times).
```

> However, it's important to note that although `DateTimeOffset` contains more information than `DateTime` by storing the offset to UTC, it isn't tied to a specific time zone, and it doesn't reflect the actual time zone to which that offset belongs. This information must be stored separately to have a full picture of the moment in time.

Reviewer 3: I don't feel this talks about why is this an issue. It is useful information and at the same time, it should go into a different segment.

Reviewer 2: Added a reference to the `Pitfalls` section, we can discuss putting everything inside this section and avoid that one entirely, let me know what do you think.

> Since the problems of `DateTime` were recognized early, a better alternative in the form of `DateTimeOffset` was introduced, and it has been the recommended variant since .NET 1.1. 
> Despite this, developers often use `DateTime` (sometimes only because it appears earlier in IntelliSense), despite its numerous limitations and potential for errors.

Reviewer: ```suggestion
```

I don't find this info particularly useful for this rspec. :/

== How to fix it

> You can refer to Microsoft Documentation to learn how to convert `DateTime` to `DateTimeOffset` succesfully. 
> Generically speaking, to migrate code to use `DateTimeOffset`, you can replace the `DateTime` variables with `DateTimeOffset` variables and use the `DateTimeOffset` properties.
> 
> Keep in mind that `DateTimeOffset` includes the time zone offset, whereas `DateTime` does not, so you may need to update any code that relies on the time zone to use the appropriate `DateTimeOffset` methods instead.

Reviewer: ```suggestion
In most cases, you can directly replace `DateTime` with `DateTimeOffset`.
For more information see DateTime(add link) and DateTimeOffset(add link) documentation.
```
They have the same properties - it's IMO a relatively easy transition.
The opposite is that it's not possible.
The should be no `TimeZone` issues as `DateTime` does not carry any info on timezones in the first place.

=== Code examples

==== Noncompliant code example

[source,csharp]
----
DateTime myDate = new DateTime(2008, 6, 19, 7, 0, 0); // Noncompliant

> myDate = DateTime.SpecifyKind(myDate, DateTimeKind.Local); 

Reviewer: I would also add here some examples with:

- `var now = DateTime.Now`
- `var nowButABitLater = DateTime.UtcNow`
- `TimeSpan diff = d - d2`

these if Im not wrong work fine if you just replace `DateTime` with `DateTimeOffset`.

Reviewer 2: `DateTime` properties are a good addition, thanks! 
Not sure about `TimeSpan` tho, the rule will raise only on `DateTime` expressions and transitioning to `DateTimeOffset` won't require any change on that side (it's working fine with both).

> myDate = DateTime.SpecifyKind(myDate, DateTimeKind.Local); 

Reviewer 3: You can use a different constructor which sets the Kind too, not to set it separately on the next line.

----

==== Compliant solution

[source,csharp]
----

> DateTimeOffset myDate = new DateTimeOffset(2008, 6, 19, 7, 0, 0, TimeSpan.FromHours(-7)); // Compliant

Reviewer 3: We need to somehow raise the point that while creating DateTime you "just" set local and all is good if you know at which location the time is reflecting then you can deduct what the timezone there was. 
With the DateTimeOffset constructor even if you want to create a local time that refers to the same location, you might need to provide different offsets based on daylight saving.
I think we should somehow mention this, because it can be that users will go and replace DateTime with DateTimeOffset, look at their current location and set an offset for the CTOR and forget that the offset might change with the daylight saving change.

----

=== Pitfalls

> Common `DateTime` pitfalls include:

Reviewer 3: These common pitfalls for DateTime make sense to me, they are valid, and at the same time I don't feel we anywhere described the reason why moving to DateTimeOffset is a good solution. What will it bring?

In my opinion, the main reason why one should use DateTimeOffset instead of DateTime is that it carries the offset and you can with 100% precision tell to which zone the time is pointing.
Which in return will even let you compare UTC vs Local times because you know the offset difference between them all the time. Even if you get a DateTimeOffset.Now and ship it to another PC in a different time zone, you still have a valid time and you can compare it with whatever you have.
While if you ship a DateTime.Now to a different machine with a different time zone you either have to somehow transfer the offset knowledge you know based on the previous PCs location or you will have wrong assumptions about the timezone.

> - When working with `DateTime` of kind `Unknown`, calling `ToUniversalTime()` presumes the `DateTime.Kind` is local and converts to UTC. When calling `ToLocalTime()`, it presumes the `DateTime.Kind` is universal and converts that to local. If you call `ToUniversalTime()`, it converts X hours in one direction, but if you call `ToLocalTime()` on it, it converts X hours in the opposite direction. Neither of these conversions can be considered correct.

Reviewer: this should be a list of things. It's too long to read and understand otherwise.

Reviewer 2: What do you mean? I can remove the second sentence that might be redundant to simplify it a bit. I'm not sure the concept is transferable to a list but I'm open to suggestions 👍
